proposals/annotation-to-mark-accessible-api-for-java.md

```diff
@@ -0,0 +1,174 @@
+so its purpose is to ensure that an API can be called from Java freely
+and to handle Kotlin features
+that were designed without attention to calling from Java
+in order to simplify development of libraries for Java and Kotlin written in Kotlin.
+
+Its behavior can be described as “annotated function is guaranteed to be available from Java by exactly the name defined either in the code or in the string parameter”; hence, the following limits are imposed:
+
+1. A `private` function cannot be marked as well as a function defined in private and local classes and objects.
+2. A `@JvmSynthetic` function cannot be marked
+3. `@JvmOverloads` is assumed
+
+Marking a `public` function without any features leading to mangling is allowed, for example, to show that the API is “Java-friendly.”
+
+Example usages:
+
+```kotlin
+class SomeThings {
+    @JvmExpose // equivalent to @JvmName("forJava")
+    internal fun forJava() {}
+}
+```
+
+Additionally, marking an `internal` function with `@JvmExpose` designates that calling it from another module in Java is **not** an error, so IDE inspections should handle it.
+
+Combination of `@JvmExpose internal` and `@JvmSynthetic public` allows creating a completely non-overlapping API for Java and Kotlin; however, it looks like abuse, so, probably, it should be an antipattern:
+
+```kotlin
+class X {
+    @JvmExpose internal fun a(consumer: java.util.function.Consumer<Int>) = consumer.accept(42)
+    @JvmSynthetic fun a(consumer: (Int) -> Unit) = consumer(42)
+}
+```
+
+### `JvmExpose` on functions with `@JvmInline` parameters or return type
+
+Functions related to `@JvmInline value class`  require special treatment since their representation in JVM differs significantly from ordinary ones.
+
+First, the value parameters and return value of a function marked with `@JvmExpose` should be boxed if their type is inline class. The reason is that inline classes are cumbersome to use from Java in their unboxed form. A mangled implementation of the function for unboxed Kotlin usage should be created as well.
+
+```kotlin
+// Example.kt
+
+@JvmInline
+value class Example(val s: String)
+
+@JvmExpose 
+fun f(x: Example): Example = TODO()
+```
+
+```java
+public static Example f-impl(java.lang.String x)
+public static Example f(Example x) {  }
+```
+
+The certain problem is that one cannot instantiate an inline class with its constructor because it is generated with `ACC_SYNTHETIC`. A solution for that could be annotating a constructor `@JvmExpose`  to have a constructor exposed by the compiler (it also will create an internal synthetic overload of it taking something like `Nothing?`). Since this requirement is unobvious, an IDE inspection must report that instances of an inline class taken as an argument of the exposed function cannot be instantiated.
+
+```kotlin
+@JvmInline
+value class Example @JvmExpose constructor(val s: String)
```

Comment:
We cannot expose default constructor, as it is used in boxing. Thus, if inline class has `init` block, it will not be executed. The only possible solution I see is to generate factory function, which is accessible only from Java.

Reply:
Like Example.of("42")? Adding it implicitly is probably a bad solution.

Reply:
Can't we do boxing by using a synthetic constructor with some flag?